File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        left, mid, right = 0, len(nums) - 2, len(nums) - 1
        result = []
        
        while right > 1:

            while left < mid:
                total = nums[left] + nums[mid] + nums[right]

                if total == 0:
                    result.append([nums[left], nums[mid], nums[right]])

                    left += 1
                    while left < mid and nums[left] == nums[left-1]:
                        left += 1
                    
                    mid -= 1
                    while mid > left and nums[mid] == nums[mid+1]:
                        mid -= 1


                elif total > 0:
                    mid -= 1

                elif total < 0:
                    left += 1

            right -= 1
            while right > 0 and nums[right] == nums[right + 1]:
                right -= 1
            
            mid = right - 1
            left = 0

        return result

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if(nums.length < 3) return result;
        Arrays.sort(nums);
        int i = 0;
        while (i < nums.length - 2) {
            if (nums[i] > 0) break;
            int j = i + 1;
            int k = nums.length - 1;

            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];

                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    while (j < k && nums[j] == nums[j+1]) {
                        j++;
                    }
                    j ++;

                    while (k > j && nums[k] == nums[k-1]) {
                        k--;
                    }
                    k--;
                } else if (sum < 0) {
                    while (j < k && nums[j] == nums[j+1]) {
                        j++;
                    }
                    j ++;
                } else if (sum > 0) {
                    while (k > j && nums[k] == nums[k-1]) {
                        k--;
                    }
                    k--;
                }
            }
            while (i < nums.length - 1 && nums[i] == nums[i+1]) {
                i++;
            }
            i++;
            
        }
        return result;
    }
}
